keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+>    { "properties": {"a": {"pattern": "2"}, "b": {"default": "2"}}}
+>  ]
+>}
+>``` 
+>
+> This says to default `b` to the same value as `a`. This cannot be expressed with a skeleton schema. In this sense, the defaults are static.
+
+Pruning and defaulting will be applied after every unstructured JSON decoding and before conversion.
+
+### Excluding values from Pruning
+
+There are cases where parts of an object are verbatim JSON, i.e. without any applied schema and especially without a complete specification which allows to apply pruning. E.g. the [`JSONSchemaProps` types in apiextensions-apiserver contain fields of type JSON](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go#L28) for defaults, examples and enums. These should never be pruned (and they are not pruned while decoding using JSON marshalling).
+
+Hence, we need a mechanism to express that in the OpenAPI validation schema (compare https://github.com/kubernetes/kubernetes/pull/64558#issuecomment-403564033).
+
+**Raw JSON Option 1:** add a format `json`, e.g.:
+
+```json
+{"properties": {"x": {"format": "json"}}}
+```
+
+In this example "x" is excluded from pruning. Note that you can still use any kind of OpenAPI validation schema constructs to restrict `"x"` further.
+
+Note that we lose expressiveness of the existing `format` strings: we either apply the format to `"json"` or any other pre-defined format. I.e. we cannot express that pruning should be disabled, but if it is an integer, it should be an int32.
+
+**Question:** this feels like a reasonable loss of expressivity. Do we accept that?
+
+**Raw JSON Option 2:** add an extension property, e.g.
+
+```json
+{"properties": {"x": {"x-no-pruning": true}}}
+```
+
+This would not lead to a loss in expressivity. It is formulated negatively intentionally to have `false` as the default with `omitempty`.
+
+As OpenAPI extensions are prefixed with `x-kubernetes` elsewhere, we probably should name it `x-kubernetes-no-pruning: true`.
+
+**We propose to follow option with the `x-kubernetes` prefix** because it is more explicit, does not reduce expressivity and does not mangle with the already very vague `format` field definition.
+
+### Nested x-no-pruning
+
+**Question:** should we support nested `x-no-pruning`, i.e. disabling pruning for a sub-object, but re-enable it something deep inside of it? E.g.
+
+```json
+{
+    "properties": {
+        "x": {
+            "x-no-pruning": true,
+            "properties": {
+		        "y": { 
+                    "x-kubernetes-no-pruning": false,
+                    "properties": { "z": {} }
+                }
+            }
+        }
+    }
+}
+```
+
+If we do, the object 
+
+```json
+{
+    "a": 1,
+    "x": {
+        "b": 2,
+        "y": {
+            "c": 3,
+            "z": 42
+        }
+    }
+}
+```
+
+is pruned to  `{"x":{"b": 2, "y":{"z": 42}}}`.
```

Comment:
The only use case I know for no pruning is an object that has a unschematized sub-tree.

So, a go type like this:

```
type FooSpec struct {
  // How many things to make
  Replicas int32
  // Object definition for things to make, can be any Kubernetes type.
  // No validation is applied, in order to support future types.
  // Just needs to be valid json/yaml.
  AnythingTemplate runtime.RawExtension
}
```
I think would be transformed into OpenAPI like this:
```
"properties":   {
    "replicas": {
        "type": "integer"
    },
    "anythingTemplate": {
        "additionalProperties": true
    }
}
```

Therefore, I think the way you express "no pruning" is by specifying that a properties has `additionalProperties: true`.  This would apply to the subtree.  No non-bool additionalProperties would be supported.


I think we disallow `additionalProperties: `

Reply:
additionalProperties only applies to one level, not to subtrees. I don't like to divert from its normal semantics. Not having `additionalProperties` is equivalent to `"additionalProperties": true`. Differentiating between those two is odd.

Reply:
Actually, I am saying something different.
I think a goal of pruning is to increase security for all types, reduce unforseen errors and provide a consistent experience for Kubernetes users.  We think that pruning should be in most cases.   But, in order to get this benefit, developers have to actively ask for pruning by laboriously adding `additionalProperties: false` to every single level.   Based on that, a likely result is that most developers don't bother, or they sometimes forget to add additionalProperties at some levels.  

Therefore, a plan is needed, I think, to encourage consistency.  Some options I see are:

1. use the OpenAPI 2.0 semantics , which are defaulting to `additionalProperties: false`.  Pro: devs have to actively opt out of pruning. Con: This would make us not be OpenAPI 3.0 standard compliant. 
2. expect that OpenAPI will be generated by tools we influence, like KubeBuilder and Operator Framework.  We will direct those frameworks to always add `additionalProperties: false` except in the case of `map` and `RawExtension` types. Pro: achieves result; con: may miss some developer workflows.
3. add a nested  `x-kubernetes-no-prune` tag. pro: devs only have to put the tag in a few places, at the root of spec and status.  con: devs still have to remember to put it in those places.